nx/c_src/pytorch.cpp

```diff
@@ -0,0 +1,327 @@
+#include <ATen/ATen.h>
+#include <iostream>
+
+#include "nx_nif_utils.hpp"
+
+ErlNifResourceType *TENSOR_TYPE;
+
+std::map<const std::string, const at::ScalarType> dtypes = {{"byte", at::kByte}, {"char", at::kChar}, {"short", at::kShort}, {"int", at::kInt}, {"long", at::kLong}, {"half", at::kHalf}, {"float", at::kFloat}, {"double", at::kDouble}};
+std::map<const std::string, const int> dtype_sizes = {{"byte", 1}, {"char", 1}, {"short", 2}, {"int", 4}, {"long", 8}, {"half", 2}, {"float", 4}, {"double", 8}};
+
+ERL_NIF_TERM create_tensor_resource(ErlNifEnv *env, at::Tensor tensor)
+{
+  ERL_NIF_TERM ret;
+  at::Tensor *tensorPtr;
+
+  tensorPtr = (at::Tensor *)enif_alloc_resource(TENSOR_TYPE, sizeof(at::Tensor));
+  if (tensorPtr == NULL)
+    return enif_make_badarg(env);
+
+  new (tensorPtr) at::Tensor(tensor.variable_data());
+
+  ret = enif_make_resource(env, tensorPtr);
+  enif_keep_resource(tensorPtr);
+  enif_release_resource(tensorPtr);
+
+  return ret;
+}
+
+at::Tensor *get_tensor(ErlNifEnv *env, ERL_NIF_TERM term)
+{
+  at::Tensor *tensorPtr;
+  if (!enif_get_resource(env, term, TENSOR_TYPE, (void **)&tensorPtr))
+  {
+    return NULL; //&(at::ones({1, 1}, at::kFloat);
+  }
+  else
+  {
+    return tensorPtr;
+  }
+}
+
+unsigned long elem_count(std::vector<int64_t> shape)
+{
+  return std::accumulate(shape.begin(), shape.end(), 1, std::multiplies<>{});
+}
+
+ERL_NIF_TERM from_blob(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+  ErlNifBinary blob;
+  std::vector<int64_t> shape;
+  std::string type;
+
+  if (!nx::nif::get_tuple(env, argv[1], shape))
+    return enif_make_badarg(env);
```

Comment:
If this fails is likely because the tensor has been deallocated. So try to raise something like `{:error, :deallocated}`. What we do in EXLA is that al NIF operations return `{:ok, value} | {:error, :deallocated}` and then we call a function called unwrap!() after every NIF call.

Something like:

```elixir
defp unwrap!({:ok, tensor}), do: tensor
defp unwrap!({:error, :deallocated}), do: raise(ArgumentError, "the tensor has been deallocated")
```

Reply:
Here it fails only if not tuple was sent.
I've changed all operations to return tuples, which go into `unwrap/1`.